File: src/Board.java
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import java.lang.Math;

public class Board extends JPanel{
	
	private final static int BOX_SIZE = SudokuSolver.BOX_SIZE;
	private final static Font baseFont = new Font("Sherif", Font.BOLD, BOX_SIZE/2);
	private final static Font solFont = new Font("Sherif", Font.PLAIN, BOX_SIZE/2);
	private final static BasicStroke thinLine = new BasicStroke(2);
	private final static BasicStroke thickLine = new BasicStroke(4);
	private int boxSelectedX = -1, boxSelectedY = -1;			// range 0-8
	private int repeatRow, repeatCol, repeatBoxX, repeatBoxY;
	private int[][] grid, sol;
	private boolean solved = false, error = false;
	private Brain brain;
	
	public void paintComponent(Graphics g) {
		super.paintComponent(g);
		this.setBackground(Color.WHITE);
		
		if(error) {
			if(repeatRow >= 0)  {
				g.setColor(Color.GRAY);
				g.fillRect(BOX_SIZE, (boxSelectedY + 1) * BOX_SIZE, 9*BOX_SIZE, BOX_SIZE);
				g.setColor(Color.CYAN);
				g.fillRect((repeatRow + 1) * BOX_SIZE, (boxSelectedY + 1) * BOX_SIZE, BOX_SIZE, BOX_SIZE);
			}
			if(repeatCol >= 0) {
				g.setColor(Color.GRAY);
				g.fillRect((boxSelectedX + 1) * BOX_SIZE, BOX_SIZE, BOX_SIZE, 9*BOX_SIZE);
				g.setColor(Color.CYAN);
				g.fillRect((boxSelectedX + 1) * BOX_SIZE, (repeatCol + 1) * BOX_SIZE, BOX_SIZE, BOX_SIZE);
			}
			if(repeatBoxX >= 0 && repeatBoxY >= 0) {
				g.setColor(Color.GRAY);
				g.fillRect((repeatBoxX/3*3+1) * BOX_SIZE, (repeatBoxY/3*3+1) * BOX_SIZE, 3*BOX_SIZE, 3*BOX_SIZE);
				g.setColor(Color.CYAN);
				g.fillRect((repeatBoxX+1) * BOX_SIZE, (repeatBoxY+1) * BOX_SIZE, BOX_SIZE, BOX_SIZE);
			}
		}
		
		// draw selected gray box
		if(boxSelectedX >= 0 && boxSelectedX < 9 && boxSelectedY >= 0 && boxSelectedY < 9) {
			g.setColor(Color.LIGHT_GRAY);
			g.fillRect((boxSelectedX + 1) * BOX_SIZE, (boxSelectedY + 1) * BOX_SIZE, BOX_SIZE, BOX_SIZE);
		}
		
		// draw board
		Graphics2D g2 = (Graphics2D) g;
		g2.setColor(Color.BLACK);
		for(int i=1; i<=10; i++) {
			if((i-1)%3 == 0) g2.setStroke(thickLine);
			else g2.setStroke(thinLine);
			g2.drawLine(i*BOX_SIZE, BOX_SIZE, i*BOX_SIZE, 10*BOX_SIZE);
			g2.drawLine(BOX_SIZE, i*BOX_SIZE, 10*BOX_SIZE, i*BOX_SIZE);
		}
		
		//draw number
		g.setColor(Color.RED);
		for(int i=0; i<9; i++) {
			for(int j=0; j<9; j++) {
				if(grid[i][j] > 0 && grid[i][j] <= 9) {
					g.setColor(Color.RED);
					g.setFont(baseFont);
					g.drawString(Integer.toString(grid[i][j]), (int)((i+1.3)*BOX_SIZE), (int)((j+1.8)*BOX_SIZE));
				}
				else if(solved) {
					g.setColor(Color.BLACK);
					g.setFont(solFont);
					g.drawString(Integer.toString(sol[i][j]), (int)((i+1.3)*BOX_SIZE), (int)((j+1.8)*BOX_SIZE));
				}
			}
		}
	}
	
	public Board(){
		grid = new int[9][9];
		brain = new Brain();
		resetError();
	}
	
	public String setBox(int num) {
		if(boxSelectedX<0 || boxSelectedX>8 || boxSelectedY<0 || boxSelectedY>8) return "Select a box";
		else {
			resetError();
			if(num == 0) {
				String s = String.format("%d removed from (%d, %d)...", grid[boxSelectedX][boxSelectedY], boxSelectedX, boxSelectedY);
				grid[boxSelectedX][boxSelectedY] = num;
				return s;
			}
			else {
				for(int i=0; i<9; i++) {
					if(i!=boxSelectedX && grid[i][boxSelectedY] == num) repeatRow = i;
					if(i!=boxSelectedY && grid[boxSelectedX][i] == num) repeatCol = i;
					int x =boxSelectedX/3*3 + (i%3), y = boxSelectedY/3*3 + i/3;
					if((x!= boxSelectedX || y!=boxSelectedY) && grid[x][y] == num) {
						repeatBoxX = x; repeatBoxY = y;
					}
				}
				String s = "";
				if(repeatRow!=-1) s+=String.format("error! row has %d alr >:(\n", num);
				if(repeatCol!=-1) s+=String.format("error! col has %d alr >:(\n", num);
				if(repeatBoxX!=-1) s+=String.format("error! box has %d alr >:(\n", num);
				if(!s.equals("")) {
					error = true;
					return s.substring(0, s.length()-1);
				}
				grid[boxSelectedX][boxSelectedY] = num;
				return String.format("%d inserted at (%d, %d)...", num, boxSelectedY+1, boxSelectedX+1);
			}
		}
	}
	
	public boolean solve() {
		resetError();
		sol = new int[9][9];
		return solved = brain.solve(grid, sol);
	}
	
	public void reset() {
		grid = new int[9][9];
		solved = false;
		boxSelectedX = -1; boxSelectedY = -1;
		resetError();
	}
	
	public void resetError() {
		repeatRow = repeatCol = repeatBoxX = repeatBoxY = -1;
		error = false;
	}
	
	public void setUnsolved() {
		solved = false;
	}
	
	public void setBoxSelected(int x, int y) {	//range 0-8
		boxSelectedX = x;
		boxSelectedY = y;
	}
	
	public void adjustBoxSelected(int dx, int dy) {
		boxSelectedX = Math.max(0, Math.min(8,boxSelectedX+dx));
		boxSelectedY = Math.max(0, Math.min(8,boxSelectedY+dy));
	}
}

File: javascript/src/board.js
class Board{
    constructor(w){
        this.size = w;
        this.boxSize = w/9;
        this.boxSelectedX = -1;
        this.boxSelectedY = -1;
        this.solved = false;
        this.brain = new Brain();
        this.grid = Array(9);
        this.sol = Array(9);
        for(let i=0; i<9; i++) {
            this.grid[i] = Array(9).fill(0);
            this.sol[i] = Array(9).fill(0);
        }
        this.resetError();
    }

    draw(ctx){
        ctx.clearRect(0,0,this.size,this.size);

        // error
        if(this.repeatRow > -1){
            ctx.fillStyle = "rgb(65,65,65)"
            ctx.fillRect(0, this.boxSelectedY * this.boxSize, 9*this.boxSize, this.boxSize);
            ctx.fillStyle = "red";
            ctx.fillRect(this.repeatRow * this.boxSize, this.boxSelectedY * this.boxSize, this.boxSize, this.boxSize);
        }
        if(this.repeatCol >= 0) {
            ctx.fillStyle = "rgb(65,65,65)"
            ctx.fillRect(this.boxSelectedX * this.boxSize, 0, this.boxSize, 9*this.boxSize);
            ctx.fillStyle = "red";
            ctx.fillRect(this.boxSelectedX * this.boxSize, this.repeatCol * this.boxSize, this.boxSize, this.boxSize);
        }
        if(this.repeatBoxX >= 0 && this.repeatBoxY >= 0) {
            ctx.fillStyle = "rgb(65,65,65)"
            ctx.fillRect((this.repeatBoxX/3>>0)*3 * this.boxSize, (this.repeatBoxY/3>>0)*3 * this.boxSize, 3*this.boxSize, 3*this.boxSize);
            ctx.fillStyle = "red";
            ctx.fillRect(this.repeatBoxX * this.boxSize, this.repeatBoxY * this.boxSize, this.boxSize, this.boxSize);
        }

        // selected box
        ctx.fillStyle = "rgb(130,130,130)"
        if(!this.solved && !this.outside()) {
            ctx.fillRect(this.boxSelectedX * this.boxSize, this.boxSelectedY * this.boxSize, this.boxSize, this.boxSize);
        }

        ctx.font = "20px Arial"
        ctx.textAlign = "center"
        for(let i=0; i<9; i++){
            for(let j=0; j<9; j++){
                if(this.grid[j][i]){
                    ctx.fillStyle = "white"
                    ctx.fillText(this.grid[j][i], (i+0.5)*this.boxSize, (j+0.6)*this.boxSize);
                }
                else if(this.solved){
                    ctx.fillStyle = "orange"
                    ctx.fillText(this.sol[j][i], (i+0.5)*this.boxSize, (j+0.6)*this.boxSize);
                }
            }
        }

        // board
        ctx.strokeStyle = "white";
        for(let i=0; i<10; i++){
            ctx.beginPath();
            if(i%3) ctx.lineWidth = 0.5
            else ctx.lineWidth = 2
            ctx.moveTo(0, i*this.boxSize);
            ctx.lineTo(this.size, i*this.boxSize);
            ctx.moveTo(i*this.boxSize, 0);
            ctx.lineTo(i*this.boxSize, this.size);
            ctx.stroke();
        }
    }

    setSelected(x, y){
        this.resetError();
        this.boxSelectedX = x/this.boxSize>>0;
        this.boxSelectedY = y/this.boxSize>>0;
    }

    changeSelected(x){
        if(this.outside()) return;
        this.resetError();
        if(x==0) this.boxSelectedX = Math.max(this.boxSelectedX-1, 0);
        if(x==1) this.boxSelectedY = Math.max(this.boxSelectedY-1, 0);
        if(x==2) this.boxSelectedX = Math.min(this.boxSelectedX+1, 8);
        if(x==3) this.boxSelectedY = Math.min(this.boxSelectedY+1, 8);
    }

    setBox(n){
        if(this.outside() || this.solved) return;
        if(!this.checkError(n))
            this.grid[this.boxSelectedY][this.boxSelectedX] = n;
    }

    checkError(n){
        this.resetError();
        if(!n) return false;
        let message = ""
        for(let i=0; i<9; i++){
            if(i!=this.boxSelectedX && this.grid[this.boxSelectedY][i] == n){
                this.repeatRow = i;
                message += "row has " + n + " alr >:(\n";
            }
            if(i!=this.boxSelectedY && this.grid[i][this.boxSelectedX] == n){
                this.repeatCol = i;
                message += "col has " + n + " alr >:(\n";
            }
            let x = (this.boxSelectedX/3>>0)*3 + (i%3), y = (this.boxSelectedY/3>>0)*3 + (i/3<<0);
            if((x!=this.boxSelectedX || y!=this.boxSelectedY) && this.grid[y][x] == n) {
                this.repeatBoxX = x; this.repeatBoxY = y;
                message += "box has " + n + " alr >:(\n";
            }
        }
        if(this.repeatRow<0 && this.repeatCol<0 && this.repeatBoxX<0) return false;
        alert(message);
        return true;
    }

    resetError(){
        this.repeatRow = -1;
        this.repeatCol = -1;
        this.repeatBoxX = -1;
        this.repeatBoxY = -1;
    }

    outside(){
        return this.boxSelectedX<0 || this.boxSelectedY<0 ||
            this.boxSelectedX>8 || this.boxSelectedY>8;
    }

    solve(){
        if(this.brain.solve(this.grid,this.sol)){
            this.solved = true;
            return true;
        }
        return false;
    }

    reset(){
        this.solved = false;
        for(let i=0; i<9; i++) {
            this.grid[i] = Array(9).fill(0);
            this.sol[i] = Array(9).fill(0);
        }
    }
}